indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java

```diff
@@ -56,33 +59,29 @@ public SupervisorManager(MetadataSupervisorManager metadataSupervisorManager)
 
   public Optional<SupervisorSpec> getSupervisorSpec(String id)
   {
-    return supervisors.get(id) == null
-           ? Optional.<SupervisorSpec>absent()
-           : Optional.fromNullable(supervisors.get(id).rhs);
+    Pair<Supervisor, SupervisorSpec> supervisor = supervisors.get(id);
+    return supervisor == null ? Optional.<SupervisorSpec>absent() : Optional.fromNullable(supervisor.rhs);
   }
 
-  public boolean hasSupervisor(String id)
-  {
-    return supervisors.containsKey(id);
-  }
-
-  public boolean createAndStartSupervisor(SupervisorSpec spec)
+  public boolean createOrUpdateAndStartSupervisor(SupervisorSpec spec)
   {
+    Preconditions.checkState(started, "SupervisorManager not started");
```

Comment:
if stop is taking a long time, it will hold the lock for a while and never set started = false. It seems reasonable to me to expect some request to come in and wait on that lock between the time that stop has acquired the lock (and began looping through its stops) and the time it sets started = false